File: Sources/HabibiFloatCore/ScreenClamp.swift
import Foundation

public enum ScreenClamp {
    public static func clampedOrigin(
        _ origin: Point2D,
        windowSize: Size2D,
        visibleFrame: Rect2D,
        margin: Double = 16
    ) -> Point2D {
        let minX = visibleFrame.x + margin
        let minY = visibleFrame.y + margin
        let maxX = visibleFrame.maxX - windowSize.width - margin
        let maxY = visibleFrame.maxY - windowSize.height - margin

        if maxX < minX || maxY < minY {
            return Point2D(
                x: visibleFrame.midX - windowSize.width / 2,
                y: visibleFrame.midY - windowSize.height / 2
            )
        }

        return Point2D(
            x: min(max(origin.x, minX), maxX),
            y: min(max(origin.y, minY), maxY)
        )
    }

    public static func defaultOrigin(windowSize: Size2D, visibleFrame: Rect2D, margin: Double = 48) -> Point2D {
        clampedOrigin(
            Point2D(
                x: visibleFrame.maxX - windowSize.width - margin,
                y: visibleFrame.maxY - windowSize.height - margin
            ),
            windowSize: windowSize,
            visibleFrame: visibleFrame,
            margin: 16
        )
    }
}

File: windows/Bubbly.Windows/Services/ScreenClamp.cs
using System;

namespace Bubbly.Windows.Services;

public sealed record PointD(double X, double Y);
public sealed record SizeD(double Width, double Height);
public sealed record RectD(double X, double Y, double Width, double Height)
{
    public double MinX => X;
    public double MaxX => X + Width;
    public double MinY => Y;
    public double MaxY => Y + Height;
    public double MidX => X + Width / 2;
    public double MidY => Y + Height / 2;
}

public static class ScreenClamp
{
    public static PointD DefaultOrigin(SizeD windowSize, RectD visibleFrame)
    {
        return new PointD(
            visibleFrame.MaxX - windowSize.Width - 42,
            visibleFrame.MinY + visibleFrame.Height * 0.62);
    }

    public static PointD ClampedOrigin(PointD origin, SizeD windowSize, RectD visibleFrame)
    {
        var minX = visibleFrame.MinX;
        var maxX = Math.Max(minX, visibleFrame.MaxX - windowSize.Width);
        var minY = visibleFrame.MinY;
        var maxY = Math.Max(minY, visibleFrame.MaxY - windowSize.Height);

        return new PointD(
            Math.Clamp(origin.X, minX, maxX),
            Math.Clamp(origin.Y, minY, maxY));
    }
}
